cross-root-aria-reflection.md

```diff
@@ -510,6 +510,53 @@ and Reflection APIs are making it possible for more complex interfaces to be acc
 with custom element and shadow DOM without needing to architect your whole implementation around 
 the intricacies of keeping ID references in a single DOM tree.
 
+### Other patterns
+
+There are many patterns where the presence of a shadow boundary will prevent otherwise default
+relationships between DOM element. Another that this API could directly benefit is a button group 
+that manages selection on those buttons similar to what we see in a collection of radio buttons (one 
+selected button) or checkboxes (multiple selected buttons). 
+
+Generally, this contract is made by having a `role="radiogroup"` or `role="group"` parent gather a 
+collection of `role="radio"` or `role="checkbox"` elements, respectively. In this case the native 
+semantics and focusability of a `<button>` element can do a lot of the heavy lifting, but if you do 
+do inside of a shadow boundary you run into some problems:
+
+```html 
+<z-button-group role="radiogroup">
+    <z-button>
+        #shadow-root
+            <button role="radio"><slot></slot></button>
+        Option 1
+    </z-button>
+    <z-button>
+        #shadow-root
+            <button role="radio"><slot></slot></button>
+        Option 2
+    </z-button>
+</z-button-group>
+```
+
+Here the `role="radio"` elements do not share a DOM tree with the `role="radiogroup"` elements which 
+breaks the contract requried to build the correct accessibility tree and pass it on to screen 
+readers. Here is another useful place for us to reflect aria attribtues from the shadow DOM into 
+the parent DOM tree:
+
+```html 
+<z-button-group role="radiogroup">
+    <z-button>
+        #shadow-root delegates="focus" reflects="role"
+            <button role="radio" reflect-role><slot></slot></button>
+        Option 1
+    </z-button>
+    <z-button>
+        #shadow-root delegates="focus" reflects="role"
+            <button role="radio" reflect-role><slot></slot></button>
+        Option 2
+    </z-button>
+</z-button-group>
+```
+
```

Comment:
This is a great point and something raised by our internal developers as well. E.g. when building custom radio buttons, you can't have a `name` that "just works" for ensuring that only one radio button with that name is selected. Part of this is solved by [focus group](https://github.com/openui/open-ui/blob/main/research/src/pages/focusgroup/focusgroup.explainer.mdx), but AFAICT it's not fully solved. (I.e. it solves keyboard events, but not the "only one radio button can be selected at a time" issue.)

@bkardell had mentioned that it may be useful to have a concept of "mixins" or "behaviors" that could apply to custom elements. (I.e. you could grab several "behaviors" from radio buttons, rather than needing to extend the entire radio button.) This definitely seems related to the idea of reflection.

Reply:
This is were we loop back to https://github.com/Westbrook/cross-root-aria-reflection/pull/3/files/c5811fa8f6c68b42ee05fd3a173e0a88671a709b#r917390572 and I need to think about how the `role="radio"` gets into the `<z-button>` element to begin with while allowing there to be ONLY one element here with an interactive role.

I _think_ this might want:
```html
    <z-button role="radio">
        #shadow-root delegates="focus role" reflects="role"
            <button auto-role reflect-role><slot></slot></button>
        Option 2
    </z-button>
```

Here we apply `role="radio"` to the root, that role is then delegated to the `<button>`, which reflects the role to the `<z-button>` element, so that it should fulfill the child relationship with the parent `role="radiogroup"` element.

If that's not proper confusing, I don't know what is. However, it seems like what we want to be possible, unless we need to wait for `mixins`/`behaviors` to make this example fully work...

Reply:
@nolanlawson is that more clear?